lib/svgo/coa.js

```diff
@@ -300,23 +301,32 @@ function optimizeFolder(config, dir, output) {
  * @param {string} output output directory
  * @return {Promise}
  */
-function processDirectory(config, dir, files, output) {
+async function processDirectory(config, dir, files, output) {
   // take only *.svg files, recursively if necessary
   var svgFilesDescriptions = getFilesDescriptions(config, dir, files, output);
-
-  return svgFilesDescriptions.length
-    ? Promise.all(
-        svgFilesDescriptions.map((fileDescription) =>
-          optimizeFile(
-            config,
-            fileDescription.inputPath,
-            fileDescription.outputPath,
-          ),
-        ),
-      )
-    : Promise.reject(
-        new Error(`No SVG files have been found in '${dir}' directory.`),
-      );
+  
+  if ( svgFilesDescriptions.length == 0 ) {
+    throw 'No SVG files have been found in "' + dir + '" directory.';
+  }
+  
+  let results = [ ],
+      cpus = os.cpus().length;
+
+  while ( svgFilesDescriptions.length > 0 ) {
+    let _svgFilesDescriptions = svgFilesDescriptions.splice(0, cpus);
+    
+    _svgFilesDescriptions = _svgFilesDescriptions.map(function(fileDescription) {
+      return optimizeFile(
+        config,
+        fileDescription.inputPath,
+        fileDescription.outputPath,
+      );   
+    });
+    
+    results = results.concat(await Promise.all(_svgFilesDescriptions));
```

Comment:
This is slower than the existing version because each "chunk" of files takes as long as the longest file. I would suggest changing this to a worker-based method, let me know if you want me to explain

Reply:
Yes, this is slower. This PR in fact run a throttle mechanism over chunks, this is the goal. Workers dont fix this, under linux the kernel has a limit over how many files can be open at same time, this limit is shared by all process of user, and another limit is shared for all process. The only way to prevent EMFILE error runing svgo over a huge number of svg is go slower and do the job in chunks.
The performance impact is minimun, runing over folder with 80k+ of svg works well.

Reply:
You're misunderstanding what I said. Right now, it goes like "10 files open - > 6 files open - > 1 file open" if one takes longer than the rest, leading to it taking longer to start the next set of files. If you instead create 10 workers, each one will continously have 1 file open.